File: PrayTimes.java
/* ----< Copyright Block >----
 *
 * Prayer Times Calculator
 *     PrayTimes.java
 *     Copyright (C) 2012 
 *     Mohsen Izadi <m.izadi@gmail.com>
 *
 * Based On:
 *     PrayTimes.js (v2.3)
 *     Copyright (C) 2007-2011 PrayTimes.org
 *     Hamid Zarrabi-Zadeh
 *
 * License: GNU LGPL v3.0
 *
 * TERMS OF USE:
 *     Permission is granted to use this code, with or 
 *     without modification, in any website or application 
 *     provided that credit is given to the original work 
 *     with a link back to PrayTimes.org.
 *
 * This program is distributed in the hope that it will 
 * be useful, but WITHOUT ANY WARRANTY. 
 *
 * PLEASE DO NOT REMOVE THIS COPYRIGHT BLOCK.
 *
 */

package org.praytimes.www;

import java.util.GregorianCalendar;
import java.util.Map;
import java.util.EnumMap;
import java.util.TimeZone;

public class PrayTimes {

	private class SunPosition {
		private final double declination;
		private final double equation;
		
		public SunPosition(double declination, double equaiton) {
			this.declination = declination;
			this.equation = equaiton;
		}
		
		public double getDeclination() {
			return declination;
		}
		
		public double getEquation() {
			return equation;
		}
	}
	
	private Params params;
	
	public PrayTimes(Params params) {
		this.params = params;
	}
	
	public Params getParams() {
		return params;
	}

	private double timeDiff(double time1, double time2) {
		return DegMath.fixHour(time2 - time1);
	}
	
	private double julian(Date date) {
		int month = date.getMonth();
		int year = date.getYear();
		if (month <= 2) {
			year--;
			month += 12;
		}
		int a = year / 100;
		int b = 2 - a + a / 4;
		return Math.floor(365.25 * (year + 4716)) + Math.floor(30.6001 * (month + 1)) +
			date.getDay() + b - 1524.5;
	}

	private void dayPortions(Map<TimePoint, Double> times) {
		for (Map.Entry<TimePoint, Double> entry: times.entrySet())
			entry.setValue(entry.getValue() / 24.0);
	}
	
	private SunPosition sunPosition(double jDate) {
		double d = jDate - 2451545.0;
		double g = DegMath.fixAngle(357.529 + 0.98560028 * d);
		double q = DegMath.fixAngle(280.459 + 0.98564736 * d);
		double l = DegMath.fixAngle(q + 1.915 * DegMath.sin(g) + 0.020 * DegMath.sin(2 * g));

		//double r = 1.00014 - 0.01671 * DegMath.cos(g) - 0.00014 * DegMath.cos(2 * g);
		double e = 23.439 - 0.00000036 * d;

		double ra = DegMath.arctan2(DegMath.cos(e) * DegMath.sin(l), DegMath.cos(l)) / 15;
		double eqt = q / 15 - DegMath.fixHour(ra);
		double decl = DegMath.arcsin(DegMath.sin(e) * DegMath.sin(l));

		return new SunPosition(decl, eqt);
	}
	
	private double midDay(double jDate, double time) {
		double eqt = sunPosition(jDate + time).getEquation();
		return DegMath.fixHour(12 - eqt);
	}

	private double sunAngleTime(Coords coords, double jDate, double angle, double time,
		boolean ccw) {
		double decl = sunPosition(jDate + time).getDeclination();
		double noon = midDay(jDate, time);
		double t = 1.0 / 15.0 * DegMath.arccos((-DegMath.sin(angle) - DegMath.sin(decl) *
			DegMath.sin(coords.getLatitude())) / (DegMath.cos(decl) * DegMath.cos(coords.getLatitude())));
		return noon + (ccw ? -t : t);
	}
	
	private double riseSetAngle(Coords coords) {
		double angle = 0.0347 * Math.sqrt(coords.getElevation()); // an approximation
		return 0.833 + angle;
	}
	
	private double asrTime(Coords coords, double jDate, double factor, double time) {
		double decl = sunPosition(jDate + time).getDeclination();
		double angle = -DegMath.arccot(factor + DegMath.tan(Math.abs(coords.getLatitude() - decl)));
		return sunAngleTime(coords, jDate, angle, time, false);
	}
	
	private void computePrayerTimes(Coords coords, double jDate, Map<TimePoint, Double> times) {
		dayPortions(times);
		for (Map.Entry<TimePoint, Double> entry: times.entrySet())
			switch (entry.getKey()) {
			case IMSAK:
				if (params.getSetting().getImsak().getType() == ImsakSetting.Type.TWILIGHT_ANGLE_BASED)
					entry.setValue(sunAngleTime(coords, jDate, params.getSetting().getImsak().asTwilightAngleBased().getTwilightAngle(),
						entry.getValue(), true));
				break;
			case FAJR:
				entry.setValue(sunAngleTime(coords, jDate, params.getSetting().getFajr().getTwilightAngle(), entry.getValue(), true));
				break;
			case SUNRISE:
				entry.setValue(sunAngleTime(coords, jDate, riseSetAngle(coords), entry.getValue(), true));
				break;
			case DHUHR:
				entry.setValue(midDay(jDate, entry.getValue()));
				break;
			case ASR:
				entry.setValue(asrTime(coords, jDate, params.getSetting().getAsr().getShadowFactor(), entry.getValue()));
				break;
			case SUNSET:
				entry.setValue(sunAngleTime(coords, jDate, riseSetAngle(coords), entry.getValue(), false));
				break;
			case MAGHRIB:
				if (params.getSetting().getMaghrib().getType() == MaghribSetting.Type.TWILIGHT_ANGLE_BASED)
					entry.setValue(sunAngleTime(coords, jDate, params.getSetting().getMaghrib().asTwilightAngleBased().getTwilightAngle(),
						entry.getValue(), false));
				break;
			case ISHA:
				if (params.getSetting().getIsha().getType() == IshaSetting.Type.TWILIGHT_ANGLE_BASED)
					entry.setValue(sunAngleTime(coords, jDate, params.getSetting().getIsha().asTwilightAngleBased().getTwilightAngle(),
						entry.getValue(), false));
				break;
			}
	}

	private double nightPortion(double angle, double night) {
		switch (params.getSetting().getHighLats()) {
		case ANGLE_BASED_METHOD:
			return night * angle / 60.0;
		case ONE_SEVENTH_METHOD:
			return night / 7.0;
		default:
			return night / 2.0;
		}
	}
	
	private double adjustHlTime(double time, double base, double angle, double night, boolean ccw) {
		double portion = nightPortion(angle, night);
		double tdiff = ccw ? timeDiff(time, base) : timeDiff(base, time);
		if (Double.isNaN(time) || tdiff > portion)
			time = base + (ccw ? -portion : portion);
		return time;
	}
	
	private void adjustHighLats(Map<TimePoint, Double> times) {
		double nightTime = timeDiff(times.get(TimePoint.SUNSET), times.get(TimePoint.SUNRISE));
		if (params.getSetting().getImsak().getType() == ImsakSetting.Type.TWILIGHT_ANGLE_BASED)
			times.put(TimePoint.IMSAK, adjustHlTime(times.get(TimePoint.IMSAK), times.get(TimePoint.SUNRISE),
				params.getSetting().getImsak().asTwilightAngleBased().getTwilightAngle(), nightTime, true));
		times.put(TimePoint.FAJR, adjustHlTime(times.get(TimePoint.FAJR), times.get(TimePoint.SUNRISE),
			params.getSetting().getFajr().getTwilightAngle(), nightTime, true));
		if (params.getSetting().getIsha().getType() == IshaSetting.Type.TWILIGHT_ANGLE_BASED)
			times.put(TimePoint.ISHA, adjustHlTime(times.get(TimePoint.ISHA), times.get(TimePoint.SUNSET),
				params.getSetting().getIsha().asTwilightAngleBased().getTwilightAngle(), nightTime, false));
		if (params.getSetting().getMaghrib().getType() == MaghribSetting.Type.TWILIGHT_ANGLE_BASED)
			times.put(TimePoint.MAGHRIB, adjustHlTime(times.get(TimePoint.MAGHRIB), times.get(TimePoint.SUNSET),
				params.getSetting().getMaghrib().asTwilightAngleBased().getTwilightAngle(), nightTime, false));
	}
	
	private void adjustTimes(Map<TimePoint, Double> times, Coords coords, double timeZone) {
		for (Map.Entry<TimePoint, Double> entry: times.entrySet())
			entry.setValue(entry.getValue() + timeZone - coords.getLongitude() / 15.0);
		if (params.getSetting().getHighLats() != HigherLatitudesSetting.NONE)
			adjustHighLats(times);
		if (params.getSetting().getImsak().getType() == ImsakSetting.Type.FAJR_BASED)
			times.put(TimePoint.IMSAK, times.get(TimePoint.FAJR) -
				params.getSetting().getImsak().asFajrBased().getMinutesBeforeFajr() / 60.0);
		if (params.getSetting().getMaghrib().getType() == MaghribSetting.Type.SUNSET_BASED)
			times.put(TimePoint.MAGHRIB, times.get(TimePoint.SUNSET) +
				params.getSetting().getMaghrib().asSunsetBased().getMinutesAfterSunset() / 60.0);
		if (params.getSetting().getIsha().getType() == IshaSetting.Type.MAGHRIB_BASED)
			times.put(TimePoint.ISHA, times.get(TimePoint.MAGHRIB) +
				params.getSetting().getIsha().asMaghribBased().getMinutesAfterMaghrib() / 60.0);
		times.put(TimePoint.DHUHR, times.get(TimePoint.DHUHR) + params.getSetting().getDhuhr().getMinutesAfterMidDay() / 60.0);
	}
	
	private void tuneTimes(Map<TimePoint, Double> times) {
		for (Map.Entry<TimePoint, Double> entry: times.entrySet())
			entry.setValue(entry.getValue() + params.getOffset(entry.getKey()) / 60.0);
	}
	
	private Map<TimePoint, Double> computeTimes(Coords coords, double jDate, double timeZone) {
		Map<TimePoint, Double> times = new EnumMap<TimePoint, Double>(TimePoint.class);
		times.put(TimePoint.IMSAK, 5.0);
		times.put(TimePoint.FAJR, 5.0);
		times.put(TimePoint.SUNRISE, 6.0);
		times.put(TimePoint.DHUHR, 12.0);
		times.put(TimePoint.ASR, 13.0);
		times.put(TimePoint.SUNSET, 18.0);
		times.put(TimePoint.MAGHRIB, 18.0);
		times.put(TimePoint.ISHA, 18.0);
		for (int i = 0; i < params.getNumIterations(); i++)
			computePrayerTimes(coords, jDate, times);
		adjustTimes(times, coords, timeZone);
		times.put(TimePoint.MIDNIGHT, params.getSetting().getMidnight() == MidnightSetting.JAFARI_METHOD ?
			times.get(TimePoint.MAGHRIB) + timeDiff(times.get(TimePoint.MAGHRIB), times.get(TimePoint.FAJR)) / 2.0 :
			times.get(TimePoint.SUNSET) + timeDiff(times.get(TimePoint.SUNSET), times.get(TimePoint.SUNRISE)) / 2.0);
		tuneTimes(times);
		return times;
	}
	
	public Map<TimePoint, DateTime> getTimes(Date date, Coords coords, TimeZone timeZone) {
		Map<TimePoint, Double> times = computeTimes(coords, julian(date) - coords.getLongitude() / (15.0 * 24.0),
			timeZone.getOffset(GregorianCalendar.AD, date.getYear(), date.getMonth() - 1, date.getDay(),
				0, 0) / (1000.0 * 60.0 * 60.0));

		Map<TimePoint, DateTime> result = new EnumMap<TimePoint, DateTime>(TimePoint.class);
		for (Map.Entry<TimePoint, Double> entry: times.entrySet()) {
			double h = entry.getValue();
			int hour = (int) Math.floor(h);
			double m = (h - hour) * 60;
			int min = (int) Math.floor(m);
			double s = (m - min) * 60;
			int sec = (int) Math.floor(s);
			int milli = (int) Math.round((s - sec) * 1000);
			result.put(entry.getKey(), new DateTime(date, hour, min, sec, milli));
		}
		return result;
	}
	public Map<TimePoint, DateTime> getTimes(Date date, Coords coords) {
		return getTimes(date, coords, TimeZone.getDefault());
	}
	public Map<TimePoint, DateTime> getTimes(Date date, Coords coords, double timeZone, boolean dst) {
		int hour = (int) Math.floor(timeZone);
		int min = (int) Math.round((timeZone - hour) * 60);
		if (dst) hour++;
		return getTimes(date, coords, TimeZone.getTimeZone(String.format("GMT%1$+d%2$02d", hour, min)));
	}
}

File: PrayTimes.js
/* ----< Copyright Block >----
 *
 * Prayer Times Calculator
 *     PrayTimes.js
 *     Copyright (C) 2015
 *     Mohsen Izadi <m.izadi@gmail.com>
 *
 * Based On:
 *     PrayTimes.js (v2.3)
 *     Copyright (C) 2007-2011 PrayTimes.org
 *     Hamid Zarrabi-Zadeh
 *
 * License: GNU LGPL v3.0
 *
 * TERMS OF USE:
 *     Permission is granted to use this code, with or
 *     without modification, in any website or application
 *     provided that credit is given to the original work
 *     with a link back to PrayTimes.org.
 *
 * This program is distributed in the hope that it will
 * be useful, but WITHOUT ANY WARRANTY.
 *
 * PLEASE DO NOT REMOVE THIS COPYRIGHT BLOCK.
 *
 */

var libpraytimes = libpraytimes || {};

libpraytimes.PrayTimes = function() {
  var SunPosition = function() {
    var create = function(declination, equaiton) {
      var that = {};

      var getDeclination = function() {
        return declination;
      };

      var getEquation = function() {
        return equaiton;
      };

      that.getDeclination = getDeclination;
      that.getEquation = getEquation;

      return that;
    };

    return {
      create: create,
    };
  }();

  var create = function(calculation) {
    var that = {};

    var timeDiff = function(time1, time2) {
      return libpraytimes.DegMath.fixHour(time2 - time1);
    };

    var julian = function(date) {
      var day = date.getUTCDate();
      var month = date.getUTCMonth() + 1;
      var year = date.getUTCFullYear();
      if (month <= 2) {
        year--;
        month += 12;
      }
      var a = Math.floor(year / 100);
      var b = 2 - a + Math.floor(a / 4);
      return Math.floor(365.25 * (year + 4716)) +
             Math.floor(30.6001 * (month + 1)) +
             day + b - 1524.5;
    };

    var dayPortions = function(times) {
      times.inmap(function(timePoint, value) {
        return value / 24;
      });
    };

    var sunPosition = function(jDate) {
      var d = jDate - 2451545.0;
      var g = libpraytimes.DegMath.fixAngle(357.529 + 0.98560028 * d);
      var q = libpraytimes.DegMath.fixAngle(280.459 + 0.98564736 * d);
      var l = libpraytimes.DegMath.fixAngle(
          q + 1.915 * libpraytimes.DegMath.sin(g) + 0.020 * libpraytimes.DegMath.sin(2 * g));

      //double r = 1.00014 - 0.01671 * libpraytimes.DegMath.cos(g) -
      //           0.00014 * libpraytimes.DegMath.cos(2 * g);
      var e = 23.439 - 0.00000036 * d;

      var ra = libpraytimes.DegMath.arctan2(libpraytimes.DegMath.cos(e) * libpraytimes.DegMath.sin(l),
                                            libpraytimes.DegMath.cos(l)) / 15;
      var eqt = q / 15 - libpraytimes.DegMath.fixHour(ra);
      var decl = libpraytimes.DegMath.arcsin(libpraytimes.DegMath.sin(e) * libpraytimes.DegMath.sin(l));

      return SunPosition.create(decl, eqt);
    };

    var midDay = function(jDate, time) {
      var eqt = sunPosition(jDate + time).getEquation();
      return libpraytimes.DegMath.fixHour(12 - eqt);
    };

    var sunAngleTime = function(coords, jDate, angle, time, ccw) {
      var decl = sunPosition(jDate + time).getDeclination();
      var noon = midDay(jDate, time);
      var t = 1 / 15 * libpraytimes.DegMath.arccos(
          (-libpraytimes.DegMath.sin(angle) - libpraytimes.DegMath.sin(decl) * libpraytimes.DegMath.sin(coords.getLatitude())) /
          (libpraytimes.DegMath.cos(decl) * libpraytimes.DegMath.cos(coords.getLatitude())));
      return noon + (ccw ? -t : t);
    };

    var riseSetAngle = function(coords) {
      var angle = 0.0347 * Math.sqrt(coords.getElevation()); // an approximation
      return 0.833 + angle;
    };

    var asrTime = function(coords, jDate, factor, time) {
      var decl = sunPosition(jDate + time).getDeclination();
      var angle = -libpraytimes.DegMath.arccot(factor + libpraytimes.DegMath.tan(Math.abs(coords.getLatitude() - decl)));
      return sunAngleTime(coords, jDate, angle, time, false);
    };

    var computePrayerTimes = function(coords, jDate, times) {
      dayPortions(times);
      times.inmap(function(timePoint, value) {
        switch (timePoint) {
        case libpraytimes.TimePoint.IMSAK:
          if (calculation.getImsak().getType() == libpraytimes.ImsakSetting.Type.TWILIGHT_ANGLE_BASED)
            return sunAngleTime(coords, jDate, calculation.getImsak().getTwilightAngle(), value, true);
          return value;
        case libpraytimes.TimePoint.FAJR:
          return sunAngleTime(coords, jDate, calculation.getFajr().getTwilightAngle(), value, true);
        case libpraytimes.TimePoint.SUNRISE:
          return sunAngleTime(coords, jDate, riseSetAngle(coords), value, true);
        case libpraytimes.TimePoint.DHUHR:
          return midDay(jDate, value);
        case libpraytimes.TimePoint.ASR:
          return asrTime(coords, jDate, calculation.getAsr().getShadowFactor(), value);
        case libpraytimes.TimePoint.SUNSET:
          return sunAngleTime(coords, jDate, riseSetAngle(coords), value, false);
        case libpraytimes.TimePoint.MAGHRIB:
          if (calculation.getMaghrib().getType() == libpraytimes.MaghribSetting.Type.TWILIGHT_ANGLE_BASED)
            return sunAngleTime(coords, jDate, calculation.getMaghrib().getTwilightAngle(), value, false);
          return value;
        case libpraytimes.TimePoint.ISHA:
          if (calculation.getIsha().getType() == libpraytimes.IshaSetting.Type.TWILIGHT_ANGLE_BASED)
            return sunAngleTime(coords, jDate, calculation.getIsha().getTwilightAngle(), value, false);
          return value;
        }
      });
    };

    var nightPortion = function(angle, night) {
      switch (calculation.getHighLats()) {
      case libpraytimes.HigherLatitudesSetting.ANGLE_BASED_METHOD:
        return night * angle / 60;
      case libpraytimes.HigherLatitudesSetting.ONE_SEVENTH_METHOD:
        return night / 7;
      default:
        return night / 2;
      }
    }

    var adjustHlTime = function(time, base, angle, night, ccw) {
      var portion = nightPortion(angle, night);
      var tdiff = ccw ? timeDiff(time, base) : timeDiff(base, time);
      if (isNaN(time) || tdiff > portion)
        time = base + (ccw ? -portion : portion);
      return time;
    };

    var adjustHighLats = function(times) {
      var nightTime = timeDiff(times.get(libpraytimes.TimePoint.SUNSET),
                               times.get(libpraytimes.TimePoint.SUNRISE));
      if (calculation.getImsak().getType() == libpraytimes.ImsakSetting.Type.TWILIGHT_ANGLE_BASED) {
        times.set(libpraytimes.TimePoint.IMSAK,
                  adjustHlTime(times.get(libpraytimes.TimePoint.IMSAK),
                               times.get(libpraytimes.TimePoint.SUNRISE),
                               calculation.getImsak().getTwilightAngle(),
                               nightTime,
                               true));
      }
      times.set(libpraytimes.TimePoint.FAJR,
                adjustHlTime(times.get(libpraytimes.TimePoint.FAJR),
                             times.get(libpraytimes.TimePoint.SUNRISE),
                             calculation.getFajr().getTwilightAngle(),
                             nightTime,
                             true));
      if (calculation.getIsha().getType() == libpraytimes.IshaSetting.Type.TWILIGHT_ANGLE_BASED) {
        times.set(libpraytimes.TimePoint.ISHA,
                  adjustHlTime(times.get(libpraytimes.TimePoint.ISHA),
                               times.get(libpraytimes.TimePoint.SUNSET),
                               calculation.getIsha().getTwilightAngle(),
                               nightTime,
                               false));
      }
      if (calculation.getMaghrib().getType() == libpraytimes.MaghribSetting.Type.TWILIGHT_ANGLE_BASED) {
        times.set(libpraytimes.TimePoint.MAGHRIB,
                  adjustHlTime(times.get(libpraytimes.TimePoint.MAGHRIB),
                               times.get(libpraytimes.TimePoint.SUNSET),
                               calculation.getMaghrib().getTwilightAngle(),
                               nightTime,
                               false));
      }
    };

    var adjustTimes = function(times, coords) {
      times.inmap(function(timePoint, value) {
        return value - coords.getLongitude() / 15;
      });
      if (calculation.getHighLats() != libpraytimes.HigherLatitudesSetting.NONE)
        adjustHighLats(times);
      if (calculation.getImsak().getType() == libpraytimes.ImsakSetting.Type.FAJR_BASED) {
        times.set(libpraytimes.TimePoint.IMSAK,
                  times.get(libpraytimes.TimePoint.FAJR) - calculation.getImsak().getMinutesBeforeFajr() / 60);
      }
      if (calculation.getMaghrib().getType() == libpraytimes.MaghribSetting.Type.SUNSET_BASED) {
        times.set(libpraytimes.TimePoint.MAGHRIB,
                  times.get(libpraytimes.TimePoint.SUNSET) + calculation.getMaghrib().getMinutesAfterSunset() / 60);
      }
      if (calculation.getIsha().getType() == libpraytimes.IshaSetting.Type.MAGHRIB_BASED) {
        times.set(libpraytimes.TimePoint.ISHA,
                  times.get(libpraytimes.TimePoint.MAGHRIB) + calculation.getIsha().getMinutesAfterMaghrib() / 60);
      }
      times.set(libpraytimes.TimePoint.DHUHR,
                times.get(libpraytimes.TimePoint.DHUHR) + calculation.getDhuhr().getMinutesAfterMidDay() / 60);
    };

    var tuneTimes = function(times) {
      times.inmap(function(timePoint, value) {
        return value + calculation.getTuning().getOffset(timePoint) / 60;
      });
    };

    var computeTimes = function(coords, jDate) {
      var times = libpraytimes.TimePointMap.create();
      times.set(libpraytimes.TimePoint.IMSAK, 5);
      times.set(libpraytimes.TimePoint.FAJR, 5);
      times.set(libpraytimes.TimePoint.SUNRISE, 6);
      times.set(libpraytimes.TimePoint.DHUHR, 12);
      times.set(libpraytimes.TimePoint.ASR, 13);
      times.set(libpraytimes.TimePoint.SUNSET, 18);
      times.set(libpraytimes.TimePoint.MAGHRIB, 18);
      times.set(libpraytimes.TimePoint.ISHA, 18);
      for (var i = 0; i < calculation.getNumIterations(); i++)
        computePrayerTimes(coords, jDate, times);
      adjustTimes(times, coords);
      times.set(libpraytimes.TimePoint.MIDNIGHT,
                calculation.getMidnight() == libpraytimes.MidnightSetting.JAFARI_METHOD
                    ? times.get(libpraytimes.TimePoint.MAGHRIB) +
                      timeDiff(times.get(libpraytimes.TimePoint.MAGHRIB), times.get(libpraytimes.TimePoint.FAJR)) / 2
                    : times.get(libpraytimes.TimePoint.SUNSET) +
                      timeDiff(times.get(libpraytimes.TimePoint.SUNSET), times.get(libpraytimes.TimePoint.SUNRISE)) / 2);
      tuneTimes(times);
      return times;
    };

    var getTimes = function(date, coords) {
      var times = computeTimes(coords, julian(date) - coords.getLongitude() / (15 * 24));

      var result = times.outmap(function(timePoint, value) {
        return new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(),
                                 date.getUTCDate(), 0, 0, 0, value * 60 * 60 * 1000));
      });
      return result;
    };

    that.getTimes = getTimes;

    return that;
  };

  return {
    create: create,
  };
}();
